app/src/main/java/ch/epfl/culturequest/ui/scan/ScanFragment.java

```diff
@@ -84,11 +94,42 @@ public void onSurfaceTextureUpdated(@NonNull SurfaceTexture surfaceTexture) {
                                         intent.putExtra("imageUri", lastlyStoredImageUri.toString());
                                         startActivity(intent);
                                         loadingAnimation.stopLoading();
+                                    })
+                                    .exceptionally(ex -> {
+                                        loadingAnimation.stopLoading();
+                                        Throwable cause = ex.getCause();
+                                        String errorMessage;
+                                        int drawableId;
+
+                                        if (cause instanceof OpenAiFailedException) {
+                                            errorMessage = "OpenAI failed to process the art.";
+                                            drawableId = R.drawable.openai_logo;
+                                        } else if (cause instanceof RecognitionFailedException) {
+                                            errorMessage = "Art recognition failed. Please try again.";
+                                            drawableId = R.drawable.image_recognition_error;
+                                        } else if (cause instanceof WikipediaDescriptionFailedException) {
+                                            errorMessage = "Failed to retrieve description from Wikipedia.";
+                                            drawableId = R.drawable.wikipedia_error;
+                                        } else {
+                                            errorMessage = "An unknown error occurred.";
+                                            drawableId = R.drawable.unknown_error;
+                                        }
```

Comment:
This if else clause is not very readable, would there be a way to make it better? Maybe do a switch case?

Reply:
Oh too bad, good to go then!